src/e2e-node/e2e.test.ts

```diff
@@ -291,3 +299,242 @@ describe.each([
     expect(data).toEqual({ arbitrary: "json data" });
   });
 });
+
+// Load environment variables from .env.local if available:
+config({ path: __dirname });
+
+type OidcIssuer = string;
+type ClientId = string;
+type ClientSecret = string;
+type RefreshToken = string;
+type Pod = string;
+type AuthDetails = [Pod, OidcIssuer, ClientId, ClientSecret, RefreshToken];
+// Instructions for obtaining these credentials can be found here:
+// https://github.com/inrupt/solid-client-authn-js/blob/1a97ef79057941d8ac4dc328fff18333eaaeb5d1/packages/node/example/bootstrappedApp/README.md
+const serversUnderTest: AuthDetails[] = [
+  // pod.inrupt.com:
+  [
+    process.env.E2E_TEST_ESS_POD!,
+    process.env.E2E_TEST_ESS_IDP_URL!,
+    process.env.E2E_TEST_ESS_CLIENT_ID!,
+    process.env.E2E_TEST_ESS_CLIENT_SECRET!,
+    process.env.E2E_TEST_ESS_REFRESH_TOKEN!,
+  ],
+  // pod-compat.inrupt.com:
+  [
+    process.env.E2E_TEST_ESS_COMPAT_POD!,
+    process.env.E2E_TEST_ESS_COMPAT_IDP_URL!,
+    process.env.E2E_TEST_ESS_COMPAT_CLIENT_ID!,
+    process.env.E2E_TEST_ESS_COMPAT_CLIENT_SECRET!,
+    process.env.E2E_TEST_ESS_COMPAT_REFRESH_TOKEN!,
+  ],
+  // inrupt.net
+  // Unfortunately we cannot authenticate against Node Solid Server yet, due to this issue:
+  // https://github.com/solid/node-solid-server/issues/1533
+  // Once that is fixed, credentials can be added here, and the other `describe()` can be removed.
+];
+
+describe.each(serversUnderTest)(
+  "Authenticated end-to-end tests against %s",
+  (rootContainer, oidcIssuer, clientId, clientSecret, refreshToken) => {
+    async function getSession() {
+      const session = new Session();
+      await session.login({
+        oidcIssuer: oidcIssuer,
+        clientId: clientId,
+        clientSecret: clientSecret,
+        refreshToken: refreshToken,
+      });
+      return session;
+    }
+
+    if (rootContainer.includes("pod-compat.inrupt.com")) {
+      // pod-compat.inrupt.com seems to be experiencing some slowdowns processing POST requests,
+      // so temporarily set the test timeout to 30s for it:
+      jest.setTimeout(30000);
+    }
+
+    it("can create, read, update and delete data", async () => {
+      const session = await getSession();
+      const arbitraryPredicate = "https://arbitrary.vocab/predicate";
+
+      let newThing = createThing({ name: "e2e-test-thing" });
+      newThing = setBoolean(newThing, arbitraryPredicate, true);
+      let newDataset = createSolidDataset();
+      newDataset = setThing(newDataset, newThing);
+
+      const datasetUrl = `${rootContainer}solid-client-tests/node/crud-dataset-${session.info.sessionId}.ttl`;
+      await saveSolidDatasetAt(datasetUrl, newDataset, {
+        fetch: session.fetch,
+      });
+
+      const firstSavedDataset = await getSolidDataset(datasetUrl, {
+        fetch: session.fetch,
+      });
+      const firstSavedThing = getThing(
+        firstSavedDataset,
+        datasetUrl + "#e2e-test-thing"
+      )!;
```

Comment:
I never noticed this type of operators, and just discovered their existence ^^. That looks convenient !

Reply:
I'd be a bit reluctant with it though, it's basically a type assertion, i.e. telling TypeScript that you know a value is not going to be `null` - an actual null check is of course safer. But a runtime error here is no problem, since we want tests to fail if this doesn't hold anyway.